test: add OpenSearch to the toolchain

QuerthDP opened the pull request:
This pull request is an addition of OpenSearch to the toolchain.

As we want to use OpenSearch database as prototype implementation of Approximate Nearest Neighbour Search, we need to be able to run OpenSearch node/cluster in Scylla testing environment.

The patch contains:

- OpenSearch cluster manager
- basic OpenSearch test

Limitations:
- having in mind all the tests use the same OpenSearch instance, we must use unique names for indices to make them work properly

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=23740

QuerthDP commented:
Initial PR review by @Jadw1 and @piodul: https://github.com/zpp-2024-vector-search/scylladb/pull/8

mykaul commented:
Can't we bring in a container?

QuerthDP commented:
We have decided to integrate OpenSearch with our test environment in a similar way as the MinIO service.
I think there is no need to run it in e.g. docker container for our purposes.
As for me it felt much natural to implement it like this.

swasik commented:
> Can't we bring in a container?

BTW, we have not found any dependency running in Docker (such as MinIO mentioned by @QuerthDP) in pylib. Do you know if there is any reason for it? Or who to ask?

mykaul commented:
> > Can't we bring in a container?
> 
> BTW, we have not found any dependency running in Docker (such as MinIO mentioned by @QuerthDP) in pylib. Do you know if there is any reason for it? Or who to ask?

In dtest, the LDAP server (I think) is in a container. I think it is substantially easier, reduces the effort of maintaining the 3rd party deps to it, etc.
I don't know why Minio was not brought as a container. @tchaikov may remember the reason.

avikivity commented on install-dependencies.sh:
```diff
@@ -165,6 +165,7 @@ declare -A pip_packages=(
     [pykmip]=""
     [universalasync]=""
     [boto3-stubs[dynamodb]]=""
```
Please fix the git author/committer email of the patches to use your scylladb email.

avikivity commented:
> Can't we bring in a container?

test.py itself runs in a container. Without nested container support, it doesn't work.

avikivity commented:
Please split off the first patch into a separate pull request, since it needs to be edited by a maintainer to link (tools/toolchain/image) to a regenerated container image.

QuerthDP commented:
Changelog:
- fixed commit author emails
- moved the installation into https://github.com/scylladb/scylladb/pull/23744

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16786/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16786/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16786/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cluster/test_opensearch
#### Build Details:
- Duration: 1 hr 5 min
- Builder: i-03972ff81b8c34431 (m5d.16xlarge)

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16788/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16788/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16788/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cluster/test_opensearch
#### Build Details:
- Duration: 1 hr 0 min
- Builder: i-096a86daaf0897ae3 (m5d.16xlarge)

QuerthDP commented:
> 🔴 CI State: FAILURE
✅ - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16788/)
✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16788/)
❌ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16788/)
The following new/updated tests ran 100 times for each mode:
🔹 cluster/test_opensearch

The test obviously fails as the OpenSearch dependency is not yet installed in the CI.
Waiting for: https://github.com/scylladb/scylladb/pull/23744

QuerthDP commented:
Changelog:
- changed OpenSearch startup to once per test.py run (similar to MinIO server)
- adjusted OpenSearch Cluster Manager to run OpenSearch in containers

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17324/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17324/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17324/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cluster/test_opensearch
#### Build Details:
- Duration: 1 hr 24 min
- Builder: spider9.cloudius-systems.com

avikivity commented on test/pylib/opensearch_cluster.py:
```diff
@@ -0,0 +1,141 @@
+            result = subprocess.run(
+                ['curl', f'http://{self.address}:{port}'],
+                capture_output=True,
+            ).returncode
```
Surely urrlib is a better fit here.

mykaul commented on test/pylib/resources/opensearch.yml:
```diff
@@ -0,0 +1,21 @@
+services:
+  opensearch:
+    image: opensearchproject/opensearch:latest
```
I think opensearchproject/opensearch:3 is enough or perhaps opensearchproject/opensearch:3.0.0 (but being a dot-zero release, might need updates from time to time)

QuerthDP commented:
Changelog:
- set OpenSearch version to 3.0.0
- changed check of service from curl command to urllib request

avikivity commented:
> > Looks reasonable, but tests fail.
> 
> Tests will fail until be add opensearch-py dependency. #23744

Ok, put it in draft until then.
